framework/src/media/media_recorder.c

```diff
@@ -0,0 +1,348 @@
+		}
+
+		sparam.sched_priority = CONFIG_MEDIA_PRIORITY;
+		if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+			goto error_thread_init;
+		}
+
+		if (pthread_attr_setstacksize(&attr, CONFIG_MEDIA_STACK_SIZE) != 0) {
+			goto error_thread_init;
+		}
+		if (pthread_create(&g_rc.pth, &attr, (pthread_startroutine_t)media_recording, NULL) != 0) {
+			goto error_thread_init;
+		}
+
+		g_rc.state = RECORD_RUNNING;
+	}
+
+	RECORD_UNLOCK();
+	return RECORD_OK;
+
+error_thread_init:
+	g_rc.thread_running = false;
+	RECORD_UNLOCK();
+	return RECORD_ERROR_THREAD_INIT;
+}
+
+record_result_t media_record_set_config(int channel, int sample_rate, int pcm_format, media_format_e format)
+{
+	RECORD_LOCK();
+
+	if (g_rc.pcmconf) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}
+
+	g_rc.pcmconf = (struct pcm_config *)malloc(sizeof(struct pcm_config));
+	if (!g_rc.pcmconf) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_MALLOC;
+	}
+
+	g_rc.pcmconf->channels = channel;
+	g_rc.pcmconf->rate = sample_rate;
+	g_rc.pcmconf->format = pcm_format;
+	g_rc.format = format;
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_record_prepare(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.pcmin) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}	
+
+	g_rc.pcmin = pcm_open(0, 0, PCM_IN, g_rc.pcmconf);
+	if (!g_rc.pcmin) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_DRIVER;
+	}
+
+	pcm_prepare(g_rc.pcmin);
+	g_rc.buffer_size = pcm_frames_to_bytes(g_rc.pcmin, pcm_get_buffer_size(g_rc.pcmin));
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+bool media_is_recording(void)
+{
+	return (g_rc.state == RECORD_RUNNING);
+}
+
+record_result_t media_stop_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state == RECORD_PAUSED) {
+		g_rc.state = RECORD_STOPPING;
+		sem_post(&g_rc.sem_pause);
+	} else if (g_rc.state != RECORD_RUNNING) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_STOPPING;
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_pause_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_RUNNING) {
+		RECORD_UNLOCK();		
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_PAUSED;
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_resume_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_PAUSED) {
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_RUNNING;
+	media_record_prepare();	
+	sem_post(&g_rc.sem_pause);
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_record_init(void)
+{
+	if (!b_started) {
+		b_started = true;
+		memset(&g_rc, 0, sizeof(g_rc));
+
+		pthread_mutexattr_init(&g_rc.mutex_attr_record);
+		pthread_mutexattr_settype(&g_rc.mutex_attr_record, PTHREAD_MUTEX_RECURSIVE);
+		pthread_mutex_init(&g_rc.mutex_record, &g_rc.mutex_attr_record);
+
+		g_rc.fd = -1;
+		g_rc.state = RECORD_AVAILABLE;
```

Comment:
when pcm_open is succeeded, this state should be set. not here.
RECORD_ERROR is correct, I think

Reply:
RECORD_AVAILABLE means the recorder is available for recording, error would indicate the need to initiliaze locks and other data structures. when pcm_open is success RUNNING should be set  because no other application can then record pcm until this record operation is stopped.

Reply:
I mean as a initlized value.
AVAILABLE means it is ready to record, in otherwords it prepared.
if RUNNING set after pcm_open but before writei or readi, there's some error situation
Assume that there's LED controller for recorder, in this case it will turn on if user call only record_prepare.
Of course, generally user will call writei or readi right after call prepare, but I think we can't pass QE test of MCD..
In this case RECORD_READY or something needed.